# A03 - Factores cualitativos

|                |                  |
:----------------|------------------|
| **Nombre**     | Omar Mendoza     |
| **Fecha**      | 03/09/2026       |
| **Expediente** | 757806           |

In [2]:
import pandas as pd
import statsmodels.api as sm

# Los datos empiezan en la columna D de la hoja
df = pd.read_excel("Libro1.xlsx", sheet_name="Hoja1", usecols="D:G")
df.shape

(20, 4)

In [3]:
df.head()

,Sabor,Tamaño,Hielo,Calificacion
0,Naranja,Chica,Si,8.5
1,Naranja,Chica,No,7.0
2,Naranja,Grande,Si,9.0
3,Naranja,Grande,No,7.5
4,Coco,Chica,Si,7.5


In [4]:
df.tail()

,Sabor,Tamaño,Hielo,Calificacion
15,Fresa-Piña,Grande,No,6.0
16,Sandia,Chica,Si,7.0
17,Sandia,Chica,No,5.0
18,Sandia,Grande,Si,7.5
19,Sandia,Grande,No,6.0


In [5]:
df.describe()

,Calificacion
count,20.000000
mean,7.600000
std,1.262829
min,5.000000
25%,7.000000
50%,7.500000
75%,8.500000
max,10.000000


In [6]:
df.describe(include="all")

,Sabor,Tamaño,Hielo,Calificacion
count,20,20,20,20.000000
unique,5,2,2,NaN
top,Naranja,Chica,Si,NaN
freq,4,10,10,NaN
mean,NaN,NaN,NaN,7.600000
std,NaN,NaN,NaN,1.262829
min,NaN,NaN,NaN,5.000000
25%,NaN,NaN,NaN,7.000000
50%,NaN,NaN,NaN,7.500000
75%,NaN,NaN,NaN,8.500000


In [7]:
df["Sabor"].unique()

<ArrowStringArray>
['Naranja', 'Coco', 'Khalua', 'Fresa-Piña', 'Sandia']
Length: 5, dtype: str

In [8]:
for col in df.columns:
    print(col, "->", df[col].unique())

Sabor -> <ArrowStringArray>
['Naranja', 'Coco', 'Khalua', 'Fresa-Piña', 'Sandia']
Length: 5, dtype: str
Tamaño -> <ArrowStringArray>
['Chica', 'Grande']
Length: 2, dtype: str
Hielo -> <ArrowStringArray>
['Si', 'No']
Length: 2, dtype: str
Calificacion -> [ 8.5  7.   9.   7.5  6.   8.  10.   5. ]


In [9]:
df.groupby("Sabor")["Calificacion"].describe()

,count,mean,std,min,25%,50%,75%,max
Sabor,,,,,,,,
Coco,4.0,7.125,0.853913,6.0,6.750,7.25,7.625,8.0
Fresa-Piña,4.0,7.625,1.376893,6.0,6.750,7.75,8.625,9.0
Khalua,4.0,8.875,0.853913,8.0,8.375,8.75,9.250,10.0
Naranja,4.0,8.000,0.912871,7.0,7.375,8.00,8.625,9.0
Sandia,4.0,6.375,1.108678,5.0,5.750,6.50,7.125,7.5


In [10]:
#cada sabor es 1 si la fila lo contiene, 0 si no
naranja = df["Sabor"].str.contains("Naranja").astype(int)
coco = df["Sabor"].str.contains("Coco").astype(int)
khalua = df["Sabor"].str.contains("Khalua").astype(int)
fresa_pina = df["Sabor"].str.contains("Fresa").astype(int)
sandia = df["Sabor"].str.contains("Sandia").astype(int)

# Tamaño: Chica = 0, Grande = 1    Hielo: No = 0, Si = 1
tamano = (df["Tamaño"] == "Grande").astype(int)
hielo = (df["Hielo"] == "Si").astype(int)

In [11]:
tabla = pd.DataFrame({
    "Naranja": naranja,
    "Coco": coco,
    "Khalua": khalua,
    "Fresa-Piña": fresa_pina,
    "Sandia": sandia,
    "Tamaño": tamano,
    "Hielo": hielo,
    "Calificacion": df["Calificacion"],
})

tabla

,Naranja,Coco,Khalua,Fresa-Piña,Sandia,Tamaño,Hielo,Calificacion
0,1,0,0,0,0,0,1,8.5
1,1,0,0,0,0,0,0,7.0
2,1,0,0,0,0,1,1,9.0
3,1,0,0,0,0,1,0,7.5
4,0,1,0,0,0,0,1,7.5
5,0,1,0,0,0,0,0,6.0
6,0,1,0,0,0,1,1,8.0
7,0,1,0,0,0,1,0,7.0
8,0,0,1,0,0,0,1,9.0
9,0,0,1,0,0,0,0,8.0


In [12]:
tabla.dtypes

Naranja           int64
Coco              int64
Khalua            int64
Fresa-Piña        int64
Sandia            int64
Tamaño            int64
Hielo             int64
Calificacion    float64
dtype: object

In [13]:
tabla.describe()

,Naranja,Coco,Khalua,Fresa-Piña,Sandia,Tamaño,Hielo,Calificacion
count,20.000000,20.000000,20.000000,20.000000,20.000000,20.000000,20.000000,20.000000
mean,0.200000,0.200000,0.200000,0.200000,0.200000,0.500000,0.500000,7.600000
std,0.410391,0.410391,0.410391,0.410391,0.410391,0.512989,0.512989,1.262829
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,5.000000
25%,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,7.000000
50%,0.000000,0.000000,0.000000,0.000000,0.000000,0.500000,0.500000,7.500000
75%,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,8.500000
max,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,10.000000


In [14]:
X = tabla.drop(columns="Calificacion")
X.insert(0, "Unos", 1)

y = tabla["Calificacion"]

X.head()

,Unos,Naranja,Coco,Khalua,Fresa-Piña,Sandia,Tamaño,Hielo
0,1,1,0,0,0,0,0,1
1,1,1,0,0,0,0,0,0
2,1,1,0,0,0,0,1,1
3,1,1,0,0,0,0,1,0
4,1,0,1,0,0,0,0,1


In [15]:
modelo = sm.OLS(y, X).fit()
print(modelo.summary())

                            OLS Regression Results                            
Dep. Variable:           Calificacion   R-squared:                       0.927
Model:                            OLS   Adj. R-squared:                  0.894
Method:                 Least Squares   F-statistic:                     27.67
Date:                Thu, 03 Sep 2026   Prob (F-statistic):           1.11e-06
Time:                        17:41:37   Log-Likelihood:                -6.3060
No. Observations:                  20   AIC:                             26.61
Df Residuals:                      13   BIC:                             33.58
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Unos           5.4583      0.133     41.111      0.0

Calificacion = 5.4583 + 1.4917*Naranja + 0.6167*Coco + 2.3667*Khalua + 1.1167*Fresa-Piña - 0.1333*Sandia + 0.5000*Tamaño + 1.6000*Hielo

In [16]:
modelo.params

Unos          5.458333
Naranja       1.491667
Coco          0.616667
Khalua        2.366667
Fresa-Piña    1.116667
Sandia       -0.133333
Tamaño        0.500000
Hielo         1.600000
dtype: float64

# Usando OneHotEncoder y OrdinalEncoder

In [17]:
from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder

# OneHotEncoder para el sabor: una columna 0/1 por cada sabor
ohe = OneHotEncoder(categories=[["Naranja", "Coco", "Khalua", "Fresa-Piña", "Sandia"]], dtype=int)
sabores = ohe.fit_transform(df[["Sabor"]]).toarray()

ohe.get_feature_names_out()

array(['Sabor_Naranja', 'Sabor_Coco', 'Sabor_Khalua', 'Sabor_Fresa-Piña',
       'Sabor_Sandia'], dtype=object)

In [18]:
# OrdinalEncoder para Tamaño y Hielo: Chica/No = 0, Grande/Si = 1
oe = OrdinalEncoder(categories=[["Chica", "Grande"], ["No", "Si"]], dtype=int)
ordinales = oe.fit_transform(df[["Tamaño", "Hielo"]])

oe.categories_

[array(['Chica', 'Grande'], dtype=object), array(['No', 'Si'], dtype=object)]

In [19]:
tabla_sk = pd.DataFrame(sabores, columns=ohe.categories_[0], index=df.index)
tabla_sk[["Tamaño", "Hielo"]] = ordinales
tabla_sk["Calificacion"] = df["Calificacion"]

tabla_sk

,Naranja,Coco,Khalua,Fresa-Piña,Sandia,Tamaño,Hielo,Calificacion
0,1,0,0,0,0,0,1,8.5
1,1,0,0,0,0,0,0,7.0
2,1,0,0,0,0,1,1,9.0
3,1,0,0,0,0,1,0,7.5
4,0,1,0,0,0,0,1,7.5
5,0,1,0,0,0,0,0,6.0
6,0,1,0,0,0,1,1,8.0
7,0,1,0,0,0,1,0,7.0
8,0,0,1,0,0,0,1,9.0
9,0,0,1,0,0,0,0,8.0


In [20]:
# X: columna de 1s, los 5 sabores, Tamaño y Hielo (sin Calificacion)
X_sk = tabla_sk.drop(columns="Calificacion")
X_sk.insert(0, "Unos", 1)

y_sk = tabla_sk["Calificacion"]

X_sk.head()

,Unos,Naranja,Coco,Khalua,Fresa-Piña,Sandia,Tamaño,Hielo
0,1,1,0,0,0,0,0,1
1,1,1,0,0,0,0,0,0
2,1,1,0,0,0,0,1,1
3,1,1,0,0,0,0,1,0
4,1,0,1,0,0,0,0,1


In [21]:
modelo_sk = sm.OLS(y_sk, X_sk).fit()
print(modelo_sk.summary())

                            OLS Regression Results                            
Dep. Variable:           Calificacion   R-squared:                       0.927
Model:                            OLS   Adj. R-squared:                  0.894
Method:                 Least Squares   F-statistic:                     27.67
Date:                Thu, 03 Sep 2026   Prob (F-statistic):           1.11e-06
Time:                        17:41:38   Log-Likelihood:                -6.3060
No. Observations:                  20   AIC:                             26.61
Df Residuals:                      13   BIC:                             33.58
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Unos           5.4583      0.133     41.111      0.0

Calificacion = 5.4583 + 1.4917*Naranja + 0.6167*Coco + 2.3667*Khalua + 1.1167*Fresa-Piña - 0.1333*Sandia + 0.5000*Tamaño + 1.6000*Hielo

In [22]:
modelo_sk.params


Unos          5.458333
Naranja       1.491667
Coco          0.616667
Khalua        2.366667
Fresa-Piña    1.116667
Sandia       -0.133333
Tamaño        0.500000
Hielo         1.600000
dtype: float64